### PR TITLE
[TEST] To manage gbs build time, add time limit to './testAll' @open sesame 11/02 14:59

### PR DESCRIPTION
**Pull request by ohsewon**
Set a time limit(10min) to prevent './testAll' from getting longer while gbs build.

Signed-off-by: sewon.oh <sewon.oh@samsung.com>

---
# [Template] PR Description

In general, github system automatically copies your commit message for your convenience.
Please remove unused part of the template after writing your own PR description with this template.
```bash
$ git commit -s filename1 filename2 ... [enter]

Summarize changes in around 50 characters or less

More detailed explanatory text, if necessary. Wrap it to about 72
characters or so. In some contexts, the first line is treated as the
subject of the commit and the rest of the text as the body. The
blank line separating the summary from the body is critical;
various tools like `log`, `shortlog` and `rebase` can get confused 
if you run the two together.

Further paragraphs come after blank lines.

**Changes proposed in this PR:**
- Bullet points are okay, too
- Typically a hyphen or asterisk is used for the bullet, preceded
  by a single space, with blank lines in between, but conventions vary here.

Resolves: #123
See also: #456, #789

**Self evaluation:**
1. Build test: [ ]Passed [ ]Failed [*]Skipped
2. Run test: [ ]Passed [ ]Failed [* ]Skipped

**How to evaluate:**
1. Describe how to evaluate in order to be reproduced by reviewer(s).

Add signed-off message automatically by running **$git commit -s ...** command.

$ git push origin <your_branch_name>
```

### CODE REVIEW

File: packaging/nnstreamer.spec [hunk 1]
```diff
@@ -40,6 +40,8 @@ BuildRequires: coregl-devel
 BuildRequires: cairo-devel
 # custom_example_opencv filter requires opencv-devel
 BuildRequires: opencv-devel
+# For './testAll.sh' time limit.
```
Comment (leemgs) [inline packaging/nnstreamer.spec]: It's good for maintenance even though base packages of 'gbs build' already includes this package.

File: packaging/nnstreamer.spec [hunk 2]
```diff
@@ -102,7 +104,22 @@ pushd build
 popd
 
 pushd tests
-./testAll.sh
+timeout=600
```
Comment (leemgs) [inline packaging/nnstreamer.spec]: Please write annotation enough why we have to set 600 seconds by default.

File: packaging/nnstreamer.spec [hunk 3]
```diff
@@ -102,7 +104,22 @@ pushd build
...
+do
+    sleep 1
+    if [[ ! $(ps | grep "$pid") ]]; then
+        break;
+    fi
+    (( timeout -= 1))
+done
+if [[ $(ps | grep "$pid") ]]; then
+    kill $pid
+    echo "GBS is stopped because timeout(5min) of './testAll.sh'"
```
Comment (leemgs) [inline packaging/nnstreamer.spec]: Incorrect grammar, 
(because of or because ...)???

File: packaging/nnstreamer.spec [hunk 4]
```diff
@@ -102,7 +104,22 @@ pushd build
...
+timeout=600
+./testAll.sh &
+pid=$!
+while ((timeout > 0));
+do
+    sleep 1
+    if [[ ! $(ps | grep "$pid") ]]; then
+        break;
+    fi
+    (( timeout -= 1))
```
Comment (leemgs) [inline packaging/nnstreamer.spec]: BTW. what is the meaning of ((....)) grammar?

File: packaging/nnstreamer.spec [hunk 5]
```diff
@@ -102,7 +104,22 @@ pushd build
...
+done
+if [[ $(ps | grep "$pid") ]]; then
+    kill $pid
+    echo "GBS is stopped because timeout(5min) of './testAll.sh'"
+    exit 124
```
Comment (leemgs) [inline packaging/nnstreamer.spec]: Please append a comment on a meaning of the return value 124.

Comment (ohsewon) [reply packaging/nnstreamer.spec]: fixed. 
link: https://unix.stackexchange.com/questions/205076/timeout-function-return-value

File: packaging/nnstreamer.spec (hunk 4 above)
Comment (ohsewon) [reply packaging/nnstreamer.spec]: It means to operate in parentheses.

File: packaging/nnstreamer.spec (hunk 3 above)
Comment (ohsewon) [reply packaging/nnstreamer.spec]: fixed

File: packaging/nnstreamer.spec [hunk 6]
```diff
@@ -102,7 +104,23 @@ pushd build
...
+timeout=600
+./testAll.sh &
+pid=$!
+while ((timeout > 0));
+do
+    sleep 1
+    if [[ ! $(ps | grep "$pid") ]]; then
+        break;
+    fi
+    (( timeout -= 1 ))
```
Comment (ohsewon) [inline packaging/nnstreamer.spec]: ((..)) means to operate in parentheses.
In this case, timeout is decreased by 1.

Comment (leemgs) [reply packaging/nnstreamer.spec]: Got it.

File: packaging/nnstreamer.spec [hunk 7]
```diff
@@ -102,7 +104,23 @@ pushd build
 popd
 
 pushd tests
-./testAll.sh
+# armv7l needs 6~7min to run 'testAll.sh', so timeout value is set 10min
```
Comment (leemgs) [inline packaging/nnstreamer.spec]: before: 
`# armv7l needs 6~7min to run 'testAll.sh', so timeout value is set 10min`

after: 
```
# The 'testAll.sh' script requires 6~7min to run armv7l  binary files in the current CI server. 
# The timeout value is  10min as a heuristic value from oure experience.
```

File: packaging/nnstreamer.spec (hunk 6 above)
Comment (leemgs) [reply packaging/nnstreamer.spec]: got it.

File: packaging/nnstreamer.spec (hunk 7 above)
Comment (ohsewon) [reply packaging/nnstreamer.spec]: fixed.

File: packaging/nnstreamer.spec [hunk 8]
```diff
@@ -102,7 +104,24 @@ pushd build
 popd
 
 pushd tests
-./testAll.sh
+# The 'testAll.sh' script requires 6~7min to run armv7l binary files in the current CI server.
+# The timeout value is 10min as a heuristic value from our experience.
+timeout=600
+./testAll.sh &
+pid=$!
+while ((timeout > 0));
```
Comment (leemgs) [inline packaging/nnstreamer.spec]: What is difference bwteen "(( ....))" and "( ...)" statement?

Comment (leemgs) [reply packaging/nnstreamer.spec]: which better between the below two coding styles?
My preference is first style such as the below `if` statement. :) 

It is a first style. 
```
while  ( .....); do
..........
done
```

It is second style. 
```
while (..........)
do
.....
done
```

File: packaging/nnstreamer.spec [hunk 9]
```diff
@@ -102,7 +104,23 @@ pushd build
 popd
 
 pushd tests
-./testAll.sh
+# The 'testAll.sh' script requires 6~7min to run armv7l binary files in the current CI server.
+# The timeout value is 10min as a heuristic value from our experience.
+timeout=600
+./testAll.sh &
+pid=$!
+while ((timeout > 0)); do
```
Comment (ohsewon) [inline packaging/nnstreamer.spec]: > What is difference bwteen "(( ....))" and "( ...)" statement?
> And, do we have to use ";" in while ((timeout > 0)); statement?
>

In this case, ( timeout > 0 ) is run 'timeout' with options '>' and '0' and return result of 'timeout' command.
(( timeout > 0 )) is comapre '$timeout' and '0'. if '$timeout' is bigger than 0 return true, else false.

File: packaging/nnstreamer.spec (hunk 8 above)
Comment (ohsewon) [reply packaging/nnstreamer.spec]: > What is difference bwteen "(( ....))" and "( ...)" statement?
> And, do we have to use ";" in while ((timeout > 0)); statement?
>

In this case, ( timeout > 0 ) is run 'timeout' with options '>' and '0' and return result of 'timeout' command.
(( timeout > 0 )) is comapre '$timeout' and '0'. if '$timeout' is bigger than 0 return true, else false.


> which better between the below two coding styles?
>

me too. updated.

File: packaging/nnstreamer.spec (hunk 9 above)
Comment (leemgs) [reply packaging/nnstreamer.spec]: got it.

**Pull request merged**